crypto/dh/dh_gen.c

```diff
@@ -41,7 +41,7 @@ int DH_generate_parameters_ex(DH *ret, int prime_len, int generator,
  * there is another special case method for the generators 2, 3 and 5.
  * for 2, p mod 24 == 11
  * for 3, p mod 12 == 5  <<<<< does not work for safe primes.
- * for 5, p mod 10 == 3 or 7
+ * for 5, p mod 20 == 3 or 7
  *
  * Thanks to Phil Karn for the pointers about the
```

Comment:
I wish I had a reference for this :(

Reply:
The only relevant link I can find to this is:
https://crypto.stackexchange.com/questions/12961/diffie-hellman-parameter-check-when-g-2-must-p-mod-24-11/12972, but it only talks about generator 2.

Reply:
So I think you want to explain that the original text the primes needed to match that, but then not all of them are safe primes, and so you need additional restrictions.

Reply:
I have currently no idea how to the proof for the p mod 10 == 3 or 7 might look like.
neither the other statements here.

Reply:
This is interesting: http://mathonline.wikidot.com/legendre-symbol-rules-for-3-p-and-6-p
They say about legende symbols (2/p) and (3/p):
(2/p) == 1 _if_ p mod 8 == 1 or 7
(2/p) == -1 _if_ p mod 8 == 3 or 5

(3/p) == 1 _if_ p mod 12 == 1 or 11
(3/p) == -1 _if_ p mod 12 ==  5 or 7

From (3/p) it is clear that 3 is always a residuum.
From (2/p) only 5 mod 8 is not possible, so only 3 mod 8,
which is 3, 11, or 19 mod 24
but since p == 11 mod 12 it follows that only 11 mod 24 is a solution.

Therefore the property that p == 11 mod 12 must already been used in the proof for those.

I am still looking for how to solve Legendre (5/p) though...

Reply:
From quadratic reciprocity law and 5 == 1 mod 4

(5/p) == (p/5)

(p/5) == p^2 % 5 = 1 _if_ p mod 5 == 1 or 4, -1 _if_ p mod 5 == 2 or 3
since p must be odd, we know (p/5) == -1 _if_ p mod 10 == 3 or 7

Reply:
could we use that as a reference: https://en.wikipedia.org/wiki/Legendre_symbol

it contains "Special formulas for the Legendre symbol" for
(2/p),  (3/p) and (5/p)

Reply:
So it seems that we want to generate for the non-quadratic residue, which has a 2q order, while what I read is that it might be better that we do quadratic residue instead, which an order of q.

Reply:
Yes, that is a good question.
I thought that we do not use the private key for more than one agreement, to mitigate
leaking one bit of the private key.
If we made sure that g is a quadratic residue, and we would still have to check that
the peer's public key is a quadratic residue, and not equal to p-1 in order to avoid
leaking a bit of the private key.

Reply:
Could we simply set bit 0 of the private key to 0, so it is a multiple of 2 ?

Reply:
We don't know what people are going to use this for. We internally call this DH, but it can be used for other things than DH too.

Reply:
I guess that if they want to use it for something else, they shouldn't be using DH parameters.

Reply:
For dh you can't fully trust the dh params, we should probably first check that
p is in fact a safe prime. With rfc 5114 params we do not have just a two subgroups
but **lots** of **small** subgroups. see the factorizations of the rfc 5114 groups here:
http://blog.intothesymmetry.com/2016/10/the-rfc-5114-saga.html

Reply:
Anyway, it is not possible to change the way how we generate the primes,
because of this:
https://github.com/openssl/openssl/blob/b60cba3c5d25dcd5cea2c97193a30b730c8a0f28/crypto/dh/dh_check.c#L143-L154

Reply:
I could clear the bit 0 in the private key at DH_generate_key
this would guarantee that the public key will always be a quadratic residue
as well as the shared key, without adding any overhead to DH_check_pub_key.
It does not add any security if p is not a safe prime of course,
which unfortunately only DH_check verifies, but **not** DH_check_params.

Reply:
done.